mock-recipe-server/testcases.py

```diff
@@ -0,0 +1,375 @@
+import json
+import random
+from urllib.parse import urlparse, urlunparse
+
+from normandy.base.utils import canonical_json_dumps
+from normandy.recipes.api.serializers import ClientSerializer
+from normandy.recipes.models import Action, Recipe
+from normandy.recipes.tests import ClientFactory, RecipeFactory, SignatureFactory
+
+
+def console_log(message, **kwargs):
+    return RecipeFactory(
+        action=Action.objects.get(name='console-log'),
+        arguments={'message': message},
+        **kwargs
+    )
+
+
+def show_heartbeat(**kwargs):
+    return RecipeFactory(
+        action=Action.objects.get(name='show-heartbeat'),
+        **kwargs
+    )
+
+
+def generate_long_message():
+    phrases = [
+        'You all know the mission and what is at stake.',
+        ('I have come to trust each of you with my life, but I have also heard murmurs of '
+         'discontent.'),
+        'I share your concerns.',
+        'We are trained for espionage.',
+        'We would be legends, but the records are sealed.',
+        'Glory in battle is not our way.',
+        'Think of our heroes: the Silent Step, who defeated a nation with a single shot.',
+        'Or the Ever Alert, who kept armies at bay with hidden facts.',
+        'These giants do not seem to give us solace here, but they are not all that we are.',
+        'Before the network, there was the fleet.',
+        'Before diplomacy, there were soldiers.',
+        'Our influence stopped the rachni, but before that, we held the line.',
+        'Our influence stopped the krogan, but before that, we held the line!',
+        'Our influence will stop Saren! In the battle today, we will hold the line!',
+    ]
+    return ' '.join([random.choice(phrases) for _ in range(1000)])
+
+
+def get_testcases():
+    """Return all defined testcases."""
+    return sorted(
+        [TestCaseClass() for TestCaseClass in TestCase.__subclasses__()],
+        key=lambda f: f.name
+    )
+
+
+class TestCase(object):
+    """
+    Configuration for a specific manual test case. Subclasses can
+    override methods on this class to customize the state of the server
+    that will be serialized, or, if necessary, the serialization itself.
+    """
+    @property
+    def name(self):
+        return self.__class__.__name__
+
+    @property
+    def description(self):
+        return '<p class="description">{}</p>'.format(self.__doc__)
+
+    def load(self):
+        """
+        Clear out all existing recipes and load this test's data in
+        its place.
+        """
+        Recipe.objects.all().delete()
```

Comment:
AFAICT we can't use a transaction here because the code loading data in the database and the code fetching from the database are in two separate containers; I get 500 errors from the service that I believe are because it can't find any recipes in the database.